### ISSUE TITLE
Upgrading chef-client on windows "Invalid Memory object"

### ISSUE THREAD

**Issue by stractenberg-sps** [chef-cookbooks/chef_client_updater#5]
### Cookbook version
1.0.2

### Chef-client version
12.19.36

### Platform Details
Server 2012 R2 (vSphere 6.5)

### Scenario:
Trying to update the Chef client and I always get the error "Invalid Memory object"

### Steps to Reproduce:
I created my own cookbook that includes: "include_recipe 'chef_client_updater::default'"

### Expected Result:
I'm expecting it to upgrade the Chef client to the latest stable version (13.0.118)

### Actual Result:

```
PS C:\Users\Administrator> chef-client
Starting Chef Client, version 12.19.36
[2017-04-13T12:24:05-04:00] INFO: *** Chef 12.19.36 ***
[2017-04-13T12:24:05-04:00] INFO: Platform: x64-mingw32
[2017-04-13T12:24:05-04:00] INFO: Chef-client pid: 3744
[2017-04-13T12:24:09-04:00] INFO: Run List is [recipe[update_chef_client]]
[2017-04-13T12:24:09-04:00] INFO: Run List expands to [update_chef_client]
[2017-04-13T12:24:09-04:00] INFO: Starting Chef Run for caradmin100
[2017-04-13T12:24:09-04:00] INFO: Running start handlers
[2017-04-13T12:24:09-04:00] INFO: Start handlers complete.
[2017-04-13T12:24:10-04:00] INFO: HTTP Request Returned 404 Not Found:
[2017-04-13T12:24:10-04:00] INFO: HTTP Request Returned 404 Not Found:
[2017-04-13T12:24:10-04:00] INFO: Error while reporting run start to Data Collector. URL: https://carchef01.edificeprod.local/organizations/spsanalytics/data-collector Exception: 404 -- 404 "Not Found
"  (This is normal if you do not have Chef Automate)
resolving cookbooks for run list: ["update_chef_client"]
[2017-04-13T12:24:10-04:00] INFO: Loading cookbooks [chef_client_updater@1.0.2, compat_resource@12.16.3, update_chef_client@0.2.6]
Synchronizing Cookbooks:
  - chef_client_updater (1.0.2)
  - compat_resource (12.16.3)
  - update_chef_client (0.2.6)
Installing Cookbook Gems:
Compiling Cookbooks...
Converging 1 resources
Recipe: chef_client_updater::default
  * chef_client_updater_updater[update chef-client] action update[2017-04-13T12:24:10-04:00] INFO: Processing chef_client_updater_updater[update chef-client] action update (chef_client_updater::defaul
t line 20)
[2017-04-13T12:24:11-04:00] WARN: Chef client is defined for forked runs. Sending TERM to parent process!


    ================================================================================
    Error executing action `update` on resource 'chef_client_updater_updater[update chef-client]'
    ================================================================================

    ArgumentError
    -------------
    Invalid Memory object

    Cookbook Trace:
    ---------------
    c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:128:in `run_post_install_action'
    c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:35:in `block (2 levels) in class_from_file'
    c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:32:in `block in class_from_file'

    Resource Declaration:
    ---------------------
    # In c:/chef/cache/cookbooks/chef_client_updater/recipes/default.rb

     20: chef_client_updater 'update chef-client' do
     21:   channel node['chef_client_updater']['channel']
     22:   version node['chef_client_updater']['version']
     23:   prevent_downgrade node['chef_client_updater']['prevent_downgrade']
     24:   post_install_action node['chef_client_updater']['post_install_action']
     25: end

    Compiled Resource:
    ------------------
    # Declared in c:/chef/cache/cookbooks/chef_client_updater/recipes/default.rb:20:in `from_file'

    chef_client_updater_updater("update chef-client") do
      action [:update]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :chef_client_updater
      cookbook_name "chef_client_updater"
      recipe_name "default"
      channel "stable"
      version "latest"
      prevent_downgrade true
      post_install_action "kill"
      exec_command "C:/opscode/chef/bin/chef-client"
      exec_args []
    end

    Platform:
    ---------
    x64-mingw32

[2017-04-13T12:24:11-04:00] INFO: Running queued delayed notifications before re-raising exception

Running handlers:
[2017-04-13T12:24:11-04:00] ERROR: Running exception handlers
Running handlers complete
[2017-04-13T12:24:11-04:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 06 seconds
[2017-04-13T12:24:11-04:00] FATAL: Stacktrace dumped to c:/chef/cache/chef-stacktrace.out
[2017-04-13T12:24:11-04:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2017-04-13T12:24:11-04:00] FATAL: ArgumentError: chef_client_updater_updater[update chef-client] (chef_client_updater::default line 20) had an error: ArgumentError: Invalid Memory object
PS C:\Users\Administrator>

Here is my stack trace:

Generated at 2017-04-13 12:24:27 -0400
ArgumentError: chef_client_updater_updater[update chef-client] (chef_client_updater::default line 20) had an error: ArgumentError: Invalid Memory object
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-process-0.8.3/lib/win32/process.rb:863:in `=='
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-process-0.8.3/lib/win32/process.rb:863:in `block in kill'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-process-0.8.3/lib/win32/process.rb:797:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-process-0.8.3/lib/win32/process.rb:797:in `kill'
c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:128:in `run_post_install_action'
c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:35:in `block (2 levels) in class_from_file'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/mixin/why_run.rb:52:in `add_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/provider.rb:176:in `converge_by'
c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:32:in `block in class_from_file'
(eval):2:in `block in action_update'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/provider.rb:361:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/provider.rb:361:in `compile_and_converge_action'
(eval):2:in `action_update'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/provider.rb:145:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource.rb:622:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:69:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:97:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:97:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:97:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:96:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:670:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:665:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:665:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:704:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:284:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application.rb:277:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application.rb:253:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application.rb:236:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:464:in `sleep_then_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:451:in `block in interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:450:in `loop'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:450:in `interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:434:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application.rb:59:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/bin/chef-client:26:in `<top (required)>'
C:/opscode/chef/bin/chef-client:68:in `load'
C:/opscode/chef/bin/chef-client:68:in `<main>'

>>>> Caused by ArgumentError: Invalid Memory object
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-process-0.8.3/lib/win32/process.rb:863:in `=='
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-process-0.8.3/lib/win32/process.rb:863:in `block in kill'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-process-0.8.3/lib/win32/process.rb:797:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-process-0.8.3/lib/win32/process.rb:797:in `kill'
c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:128:in `run_post_install_action'
c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:35:in `block (2 levels) in class_from_file'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/mixin/why_run.rb:52:in `add_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/provider.rb:176:in `converge_by'
c:/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:32:in `block in class_from_file'
(eval):2:in `block in action_update'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/provider.rb:361:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/provider.rb:361:in `compile_and_converge_action'
(eval):2:in `action_update'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/provider.rb:145:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource.rb:622:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:69:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:97:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:97:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:97:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/runner.rb:96:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:670:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:665:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:665:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:704:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/client.rb:284:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application.rb:277:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application.rb:253:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application.rb:236:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:464:in `sleep_then_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:451:in `block in interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:450:in `loop'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:450:in `interval_run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application/client.rb:434:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/lib/chef/application.rb:59:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36-universal-mingw32/bin/chef-client:26:in `<top (required)>'
C:/opscode/chef/bin/chef-client:68:in `load'
C:/opscode/chef/bin/chef-client:68:in `<main>'
```

**Comment by stractenberg-sps** [comment]
the formatting of the text above didn't come out as expected. If there's a trick, please let me know. Thanks.

**Comment by tas50** [comment]
This is actually a dupe of #2, which I need to set aside some time to fix. I promise to get this one resolved asap

**Comment by skares** [comment]
Hi, we have this issue even with the latest CookBook Version 1.1.1
All Windows 2012/2016 Virtual Machines have these issue. 

Recipe: chef_client_updater::default
  * chef_client_updater_updater[update chef-client] action update

    ================================================================================
    Error executing action `update` on resource 'chef_client_updater_updater[update chef-client]'
    ================================================================================

    ArgumentError
    -------------
    Invalid Memory object

    Resource Declaration:
    ---------------------
    # In C:/chef/cache/cookbooks/chef_client_updater/recipes/default.rb

     20: chef_client_updater 'update chef-client' do
     21:   channel node['chef_client_updater']['channel']
     22:   version node['chef_client_updater']['version']
     23:   prevent_downgrade node['chef_client_updater']['prevent_downgrade']
     24:   post_install_action node['chef_client_updater']['post_install_action']
     25:   download_url_override node['chef_client_updater']['download_url_override'] if node['chef_client_updater']['do
wnload_url_override']
     26:   checksum node['chef_client_updater']['checksum'] if node['chef_client_updater']['checksum']
     27: end

    Compiled Resource:
    ------------------
    # Declared in C:/chef/cache/cookbooks/chef_client_updater/recipes/default.rb:20:in `from_file'

    chef_client_updater_updater("update chef-client") do
      action [:update]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :chef_client_updater
      cookbook_name "chef_client_updater"
      recipe_name "default"
      channel "stable"
      version "latest"
      prevent_downgrade true
      post_install_action "kill"
      exec_command "C:/opscode/chef/bin/chef-client"
      exec_args []
    end

    Platform:
    ---------
    x64-mingw32

**Comment by david-in-perth** [comment]
@skares : I think @tas50 closed this one as a duplicate of #2, so you should probably follow that issue instead.
